Force Commit Consuming Segments Before Moved during Table Rebalance

J-HowHuang opened the pull request:
## Description

This PR improves the efficiency of table rebalance operations for REALTIME tables that have consuming segments. Traditionally, when a consuming segment is moved to a different server during rebalance, the new server has to re-consume from the stream, which is resource-intensive and time-consuming. By force-committing any consuming segment before it is moved, the segment transitions to ONLINE state and becomes available for direct download (or peer-download if allowed) by new servers, avoiding unnecessary re-consumption.


## Key Changes

- Add a new boolean rebalance parameter `forceCommitBeforeMoved`
- When `downtime=true` and `forceCommitBeforeMoved=true`
    - derive all consuming segments to be moved by comparing current assignment and target assignment
    - Send force commit of all these segments to the servers via `PinotLLCRealtimeSegmentManager.forceCommit`
    - Wait for these segments to have `DONE` in their ZK metadata `status` (so that it's downloadable).
    - Re-calculate the target assignment as the IS has changed
    - Update the IS to the new target assignment
- When `downtime=false` and `forceCommitBeforeMoved=true`
    - Before each time the "next assignment" is updated to the IS, derive all consuming segments to be moved in this new assignment.
    - Send force commit of all these segments to the servers via `PinotLLCRealtimeSegmentManager.forceCommit`
    - Wait for these segments to have `DONE` in their ZK metadata `status` (so that it's downloadable).
    - Re-calculate the target assignment as the IS has changed
    - Re-calculate the next assignment again, update to the IS.
- Add new fields `isForceCommittingConsumingSegments` and `totalSegmentsForceCommitted` in `RebalanceProgressStats`
    - `isForceCommittingConsumingSegments` shows that the rebalance is currently waiting for the force committing segments, just so when commits take longer time, the user can distinguish that it's due to committing instead of data moving.
    - We're not tracking how many "expected" segments to force commit here, becasue the commit could also happen on its own, in which case `totalSegmentsForceCommitted` would not match and could seem incomplete, causing confusions.
    - Current step shows how many consuming segments has been force committed in this step, overall shows how many in total. Both current step and overall shows the same value of `isForceCommittingConsumingSegments`.
- If a force commit fails, the rebalancer still proceed to rebalance (update the IS and so on), in which case servers may need to catch up a bigger offsets of `CONSUMING` segments.

## UI Change

![image](https://github.com/user-attachments/assets/bd902d13-e3c3-42eb-bb0c-57a5d43c91b8)

![image](https://github.com/user-attachments/assets/bca7a91a-445f-413b-baa1-45401fb0ea76)




## Notes

### The reason not to force commit once before the rebalance

We have to do force commit every step, instead of once before the step while loops. 

Regular realtime segments create themselves in a way that doesn't need to rebalance, i.e. after force commit on each consuming segments, there will be no new consuming segment that requires a rebalance anymore. 

However, segments with strict replica group strategy would not be created at the "right" servers. In this case, even if we force committed all consuming segments at the beginning, there will still be new consuming segments created during the rebalance that require a move. They can still accumulate a huge amount of offsets before they're actually moved.

Therefore, we force commit consuming segments right before they're actually moved (i.e. commit those to be moved in each current step) to minimize the offsets that servers would need to catch up with.

### The reason to wait until it's `DONE` in the ZK metadata status

Note that we check the `status` to be `DONE` instead of checking a segment becomes `ONLINE` in EV. Becasue for pauseless enabled tables, becoming `ONLINE` in EV may meanwhile have a `status` of `COMMITTING`, which is not enough to guarantee the segment to be downloadable on the new servers. See example:

```
# Example: Segment is ONLINE in EV but COMMITTING in ZK metadata

Suppose we have a REALTIME segment `myTable__0__0__20250613T1700Z`:

# After the commit (Server_1 is the committer)
# ExternalView (EV)
"myTable__0__0__20250613T1700Z": {
  "Server_1": "CONSUMING"
  "Server_2": "ONLINE"
}
# IdealState (IS)
"myTable__0__0__20250613T1700Z": {
  "Server_1": "ONLINE"
  "Server_2": "ONLINE"
}
# ZK Segment Metadata
{
  "segmentName": "myTable__0__0__20250613T1700Z",
  "status": "COMMITTING"
}

# If rebalancer proceed now because the segment becomes ONLINE in EV
# Rebalance moves this segment to Server_1, Server_3
"myTable__0__0__20250613T1700Z": {
  "Server_1": "CONSUMING",
  "Server_2": "DROPPED", <-- receieved ONLINE -> DROPPED transition
  "Server_3": "OFFLINE"  <-- because no ONLINE peer to download, stuck at OFFLINE
}
# After rebalance, IS:
"myTable__0__0__20250613T1700Z": {
  "Server_1": "ONLINE",
  "Server_3": "ONLINE"
}
# ZK Segment Metadata (unchanged):
{
  "segmentName": "myTable__0__0__20250613T1700Z",
  "status": "COMMITTING"
}

# The above case could possibly cause Server_3 to get timeout error, turn the segment into a bad state

# Even if there are some ONLINE peers available:
# Rebalance moves this segment to Server_2, Server_3
"myTable__0__0__20250613T1700Z": {
  "Server_1": "CONSUMING",
  "Server_2": "ONLINE",
  "Server_3": "ONLINE"  <-- successfully peer download from Server_2
}
# After rebalance, IS:
"myTable__0__0__20250613T1700Z": {
  "Server_2": "ONLINE",
  "Server_3": "ONLINE"
}
# ZK Segment Metadata (unchanged):
{
  "segmentName": "myTable__0__0__20250613T1700Z",
  "status": "COMMITTING"
}

# The rebalancer still have to wait until EV-IS converged (effectively having status `DONE` in ZK metadata) to move to the next step

```

## Manual Test

### Test forceCommit

Using quick start, rebalance from server 1 & 2 to server 3 & 4
Before:
![image](https://github.com/user-attachments/assets/7696e337-9304-4af9-aeb9-4fa9aabade13)
After:
![image](https://github.com/user-attachments/assets/936367ad-172d-4749-90ec-e656c71c2a12)
```
{
  "status": "DONE",
  "startTimeMs": 1749839509253,
  "timeToFinishInSeconds": 12,
  "completionStatusMsg": "Finished rebalancing with minAvailableReplicas: 1, enableStrictReplicaGroup: false, bestEfforts: false in 12207 ms.",
  "rebalanceProgressStatsOverall": {
    "totalSegmentsToBeAdded": 20,
    "totalSegmentsToBeDeleted": 20,
    "totalRemainingSegmentsToBeAdded": 0,
    "totalRemainingSegmentsToBeDeleted": 0,
    "totalRemainingSegmentsToConverge": 0,
    "totalCarryOverSegmentsToBeAdded": 0,
    "totalCarryOverSegmentsToBeDeleted": 0,
    "totalUniqueNewUntrackedSegmentsDuringRebalance": 0,
    "isForceCommittingConsumingSegments": false,
    "totalSegmentsForceCommitted": 10,
    "percentageRemainingSegmentsToBeAdded": 0,
    "percentageRemainingSegmentsToBeDeleted": 0,
    "estimatedTimeToCompleteAddsInSeconds": 0,
    "estimatedTimeToCompleteDeletesInSeconds": 0,
    "averageSegmentSizeInBytes": -1,
    "totalEstimatedDataToBeMovedInBytes": -1,
    "startTimeMs": 1749839509253
  },
  "rebalanceProgressStatsCurrentStep": {
    "totalSegmentsToBeAdded": 0,
    "totalSegmentsToBeDeleted": 0,
    "totalRemainingSegmentsToBeAdded": 0,
    "totalRemainingSegmentsToBeDeleted": 0,
    "totalRemainingSegmentsToConverge": 0,
    "totalCarryOverSegmentsToBeAdded": 0,
    "totalCarryOverSegmentsToBeDeleted": 0,
    "totalUniqueNewUntrackedSegmentsDuringRebalance": 0,
    "isForceCommittingConsumingSegments": false,
    "totalSegmentsForceCommitted": 0,
    "percentageRemainingSegmentsToBeAdded": 0,
    "percentageRemainingSegmentsToBeDeleted": 0,
    "estimatedTimeToCompleteAddsInSeconds": 0,
    "estimatedTimeToCompleteDeletesInSeconds": 0,
    "averageSegmentSizeInBytes": 0,
    "totalEstimatedDataToBeMovedInBytes": 0,
    "startTimeMs": 0
  }
...
}
```

### Test forceCommit timeout
Set `forceCommitBatchStatusCheckTimeoutMs` to `3000`, add a 5000ms delay in realtime segment commit process, rebalance:
```
{
  "status": "FAILED",
  "startTimeMs": 1750285768012,
  "timeToFinishInSeconds": 5,
  "completionStatusMsg": "Caught exception while waiting for consuming segments to force commit, aborting the rebalance: Exception occurred while waiting for the forceCommit of segments: [airlineStats__8__0__20250618T2228Z, airlineStats__2__0__20250618T2228Z, airlineStats__3__0__20250618T2228Z, airlineStats__1__0__20250618T2228Z, airlineStats__0__0__20250618T2228Z, airlineStats__5__0__20250618T2228Z, airlineStats__4__0__20250618T2228Z, airlineStats__6__0__20250618T2228Z, airlineStats__7__0__20250618T2228Z, airlineStats__9__0__20250618T2228Z], attempt count: 1, segmentsYetToBeCommitted: [airlineStats__8__0__20250618T2228Z, airlineStats__2__0__20250618T2228Z, airlineStats__3__0__20250618T2228Z, airlineStats__1__0__20250618T2228Z, airlineStats__0__0__20250618T2228Z, airlineStats__5__0__20250618T2228Z, airlineStats__4__0__20250618T2228Z, airlineStats__6__0__20250618T2228Z, airlineStats__7__0__20250618T2228Z, airlineStats__9__0__20250618T2228Z]",
  "rebalanceProgressStatsOverall": {
    ...
}
```

somandal commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java:
```diff
@@ -665,6 +665,8 @@ public RebalanceResult rebalance(
       @ApiParam(value = "Whether to update segment target tier as part of the rebalance") @DefaultValue("false")
       @QueryParam("updateTargetTier") boolean updateTargetTier,
+      @ApiParam(value = "Do force commit on consuming segments before they are rebalanced") @DefaultValue("false")
+      @QueryParam("forceCommitBeforeMoved") boolean forceCommitBeforeMoved,
```
we can perhaps simplify it even more? `forceCommit`?

somandal replied:
I'm okay with `forceCommitConsumingSegments` as well, but definitely rename both the parameter name and the internal variable name. let's always keep these consistent as far as possible

yashmayya replied:
+1 for renaming both the parameter name and all the internal names. `forceCommit` also seems okay as long as the description for the parameter is good. I think `forceCommitConsumingSegments` might better convey the intent to users who aren't already familiar with what force commits are in the context of realtime tables and consuming segments.

J-HowHuang replied:
Renamed to `forceCommit`

somandal commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -2185,7 +2185,7 @@ private void processBatchesSequentially(List<Set<String>> segmentBatchList, Stri
   }
 
-  private void waitUntilPrevBatchIsComplete(String tableNameWithType, Set<String> segmentBatchToCommit,
+  public void waitUntilPrevBatchIsComplete(String tableNameWithType, Set<String> segmentBatchToCommit,
```
this feels like a very internal implementation detail on how forceCommit works in batches internally to me and IMO shouldn't be exposed as public. Consider creating a utility function with an appropriate name in a common location and perhaps calling that in both code paths instead. Or consider utilizing the tracking method of the jobs metadata in ZK (similar to how forceCommit job status can be fetched) instead

J-HowHuang replied:
https://github.com/apache/pinot/blob/f81b03c20493c9dd8ed818c1706262eb2dd64797/pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotRealtimeTableResource.java#L255-L263
Also the controller API uses the method `getSegmentsYetToBeCommitted` too to track the forceCommit status, which is why I think it's okay to use it in https://github.com/apache/pinot/commit/b912b6a47af5a72f9c11fc9ef62ae0d67b8fad52 (the duplicate code is of <10 lines, which is only a general check and retry and I think it's acceptable to be duplicated).

Given the same reason (external use of method in `PinotLLCRealtimeSegmentManager`) I think it's also acceptable to make `waitUntilPrevBatchIsComplete` public, as my original approach. Perhaps with a change of its name to `waitUntilSegmentsAreComplete` to make the intent generic.

I prefer not to make an util for this, as I don't think this will be common enough to be so. 

@somandal What's your opinion on this?

somandal replied:
Usually it's better to have a utility if there are at least 2 code paths that use it. We often do that even if it is not wider used. Main issue with duplication is the maintenance overhead of fixing up the code in 2 or more places instead of one when making bug fixes or changes. It is easy to miss and this leads to accumulating more bugs over time.

I'm okay with making it public but prefer a rename for sure if going with that route. The name should include something about forceCommit in it, otherwise it is hard to make sense of what you're waiting for 😅 (try to think in terms of code readability without having to jump to that function etc)

J-HowHuang replied:
Renamed it to 
https://github.com/apache/pinot/blob/7d96ac40ef19c2477d4d2ed0c1007a3998e62367/pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java#L2181-L2182

somandal commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:
```diff
@@ -231,14 +237,18 @@ private RebalanceResult doRebalance(TableConfig tableConfig, RebalanceConfig reb
         minReplicasToKeepUpForNoDowntime, enableStrictReplicaGroup, lowDiskMode, bestEfforts, batchSizePerServer,
-        externalViewCheckIntervalInMs, externalViewStabilizationTimeoutInMs, minimizeDataMovement);
+        externalViewCheckIntervalInMs, externalViewStabilizationTimeoutInMs, minimizeDataMovement,
+        forceCommitBeforeMoved);
```
let's add validation that this is not enabled for `OFFLINE` or internally set this to `false` for `OFFLINE` tables with a warning log. we don't do validations for `includeConsuming` either, so I'm okay with the log message and internally setting this to false if that's better

somandal replied:
pre-check is overkill IMO as this flag should be a no-op for OFFLINE tables even if enabled. I think just adding a warning log and toggling this to false is good enough. we don't warn for includeConsuming in pre-checks either

J-HowHuang replied:
reverted the pre-check part change

somandal commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:
```diff
@@ -1970,4 +2024,50 @@ static class SingleSegmentAssignment {
+      @Nullable Set<String> segmentsToCommit, Logger tableRebalanceLogger) {
+    if (_pinotLLCRealtimeSegmentManager != null) {
+      ForceCommitBatchConfig forceCommitBatchConfig =
+          ForceCommitBatchConfig.of(Integer.MAX_VALUE, 5, 180);
```
I kind of think we should expose these as rebalance config options in case users want to provide their own values for whatever reason as limitations on how many to forceCommit at a time etc. Can choose some good defaults and +1 to Yash's comment on using constants

J-HowHuang replied:
Absolutely. Besides this, let's also decide whether it's acceptable to keep rebalancing even the force commit timeout (for ailed to finish) in the other thread before I make changes to expose the configs?

somandal reviewed (commented):
Mostly just wanted to leave a reminder to also update the UI to add this new option. Left a few other comments but haven't done a thorough review.

J-HowHuang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:
```diff
@@ -1970,4 +2024,50 @@ static class SingleSegmentAssignment {
+        // happens only after the committing segment status is set to DONE.
+        _pinotLLCRealtimeSegmentManager.waitUntilPrevBatchIsComplete(tableNameWithType, segmentsToCommit,
+            forceCommitBatchConfig);
+      } catch (Exception e) {
+        tableRebalanceLogger.warn("Failed to wait for previous batch to complete", e);
+      }
```
My initiative was that this flag tries force commit to reduce the burden from the newly spinning up servers, so it's more like an add-on. If this add-on doesn't work the whole rebalance would just be a regular rebalance.

Your point is also valid too, but I'm not sure how much the rebalance users want this force commit to be enforced. @somandal any thoughts on this?

somandal replied:
yes the rebalance should be marked as failed if this happens for now. we can revisit this later if we actually find cases where it is not useful to fail the rebalance. in the worst case the user can retry the rebalance with forceCommit=false

yashmayya replied:
> My initiative was that this flag tries force commit to reduce the burden from the newly spinning up servers, so it's more like an add-on. If this add-on doesn't work the whole rebalance would just be a regular rebalance.

There could potentially be use cases where the consumption offset lag introduced by rebalancing consuming segments might not be acceptable and the force commit config would be used to prevent that. Ignoring force commit failures / timeouts and proceeding with the rebalance could violate data freshness SLAs in such cases.

J-HowHuang replied:
Done. Return rebalance failure now if exception is thrown while waiting for force commit

J-HowHuang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:
```diff
@@ -508,66 +531,100 @@ private RebalanceResult doRebalance(TableConfig tableConfig, RebalanceConfig reb
+        }
+
+        // Step 2: Handle force commit if flag is set, then recalculate if force commit occurred
+        if (shouldForceCommit) {
+          nextAssignment =
+              getNextAssignment(currentAssignment, targetAssignment, minAvailableReplicas, enableStrictReplicaGroup,
+                  lowDiskMode, batchSizePerServer, segmentPartitionIdMap, partitionIdFetcher, tableRebalanceLogger);
+          Set<String> consumingSegmentsToMoveNext = getMovingConsumingSegments(currentAssignment, nextAssignment);
+
+          if (!consumingSegmentsToMoveNext.isEmpty()) {
```
Before the batching was introduced, yes. With batching, it's not guaranteed to have all segments to be moved in each step, so there might be a consuming segment not moving at all in the first step.

J-HowHuang replied:
Right. Unless the server instances change and IS update happens outside of the `TableRebalancer`, then the moved partition may need to be moved again

J-HowHuang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:
```diff
@@ -473,6 +492,10 @@ private RebalanceResult doRebalance(TableConfig tableConfig, RebalanceConfig reb
+            isStrictRealtimeSegmentAssignment);
     // We repeat the following steps until the target assignment is reached:
     // 1. Wait for ExternalView to converge with the IdealState. Fail the rebalance if it doesn't converge within the
     //    timeout.
```
Also added relevant texts for batch size per server. 
cc @somandal

J-HowHuang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:
```diff
@@ -1970,4 +2024,50 @@ static class SingleSegmentAssignment {
+      @Nullable Set<String> segmentsToCommit, Logger tableRebalanceLogger) {
+    if (_pinotLLCRealtimeSegmentManager != null) {
+      ForceCommitBatchConfig forceCommitBatchConfig =
+          ForceCommitBatchConfig.of(Integer.MAX_VALUE, 5, 180);
```
I'll be doing this. Though, I just realized the unit here is different to the parameter of the force commit API, which is in second, while the unit in ForceCommitBatchConfig is millisecond. Surprisingly the tests could pass with a 180ms timeout...

J-HowHuang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:
```diff
@@ -508,66 +531,100 @@ private RebalanceResult doRebalance(TableConfig tableConfig, RebalanceConfig reb
-              break;
+      Map<String, Map<String, String>> nextAssignment;
+      boolean needsRecalculation;
+      boolean shouldForceCommit = forceCommitBeforeMoved;
```
Incorrect, the branch of `downtime=true` returns in its own branch and would not ever get to the loop

J-HowHuang commented:
Thanks for the reviews @yashmayya and @somandal , most of the comments are addressed from [e3b7893](https://github.com/apache/pinot/pull/16096/commits/e3b789364a626d3813d8890d34dc6bd7cb4628be) and on.

Discussions:
1. Should we fail the rebalance job if force commit throws exception?
2. Right now I expose force commit batch size and timeout to the UI, keeping force commit check interval to the API (as other similar parameters), and default the values to what they are in the forceCommit API. Is this good?
3. What should we name the `forceCommitBeforeMoved` parameter in the API? I feel like `forceCommit` lacks its context a bit, but not strongly against it though.

Also, updated the UI and added preview in the PR description.

somandal commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/DefaultRebalancePreChecker.java:
```diff
@@ -410,6 +410,11 @@ private RebalancePreCheckerResult checkRebalanceConfig(RebalanceConfig rebalance
 
+    if (tableConfig.getTableType() == TableType.OFFLINE && rebalanceConfig.isForceCommitBeforeMoved()) {
+      pass = false;
+      warnings.add("forceCommitBeforeMoved is set for OFFLINE table, which will be ignored.");
```
let's rename this to match the parameter name

somandal replied:
though, I think adding this to pre-checks is probably overkill. why not just set it to `false` internally and instead add a warning log. we don't have a similar pre-check as this for `includeConsuming` either, do we?

yashmayya replied:
Yeah +1, I'm not sure pre-checks is the best place to put this. I'd advocate for either logging a warning and overriding it to `false` internally or returning an error back to the user indicating that the option can't be used with `OFFLINE` tables.

somandal commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/RebalanceConfig.java:
```diff
@@ -136,6 +136,22 @@ public class RebalanceConfig {
   @ApiModelProperty(example = "300000")
   private long _retryInitialDelayInMs = 300000L;
 
+  @JsonProperty("forceCommitBeforeMoved")
+  @ApiModelProperty(example = "false")
+  private boolean _forceCommitBeforeMoved = false;
```
let's rename this to be consistent with the API parameter

somandal commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:
```diff
@@ -1970,4 +2024,50 @@ static class SingleSegmentAssignment {
+      String segmentName = entry.getKey();
+      Map<String, String> currentInstanceStateMap = entry.getValue();
+      Map<String, String> targetInstanceStateMap = targetAssignment.get(segmentName);
+      if (targetInstanceStateMap != null && targetInstanceStateMap.values().stream()
+          .noneMatch(state -> state.equals(SegmentStateModel.ONLINE)) && targetInstanceStateMap.values().stream()
+          .anyMatch(state -> state.equals(SegmentStateModel.CONSUMING))) {
```
let's keep this consistent everywhere. I believe one of the segment assignment utils also does the check in the same way. If we want to simplify, I recommend opening a new PR with that change and make it everywhere at one go and test it out

somandal commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:
```diff
@@ -1970,4 +2024,50 @@ static class SingleSegmentAssignment {
+      } catch (Exception e) {
+        tableRebalanceLogger.warn("Failed to wait for previous batch to complete", e);
+      }
+    } else {
+      tableRebalanceLogger.warn(
+          "PinotLLCRealtimeSegmentManager is not initialized, cannot force commit consuming segments");
+    }
```
we have a ticket to move the command to using the controller API. Perhaps that change can be made and then this code cleaned as part of that

yashmayya replied:
> Unless the PinotHelixResourceManager have a null _pinotLLCRealtimeSegmentManager, which is unlikely?

Yeah, this shouldn't be possible. 

> we have a ticket to move the command to using the controller API. Perhaps that change can be made and then this code cleaned as part of that

Yeah sounds good, if we already have a ticket for that we can leave this part as is for now.

somandal commented on pinot-controller/src/main/resources/app/components/Homepage/Operations/RebalanceServer/RebalanceServerOptions.ts:
```diff
@@ -174,5 +174,35 @@ export const rebalanceServerOptions: RebalanceServerOption[] = [
         "markWithWarningIcon": false
+    },
+    {
+        "name": "forceCommitBeforeMoved",
```
let's use the updated name here and in the `label`

somandal commented on pinot-controller/src/main/resources/app/components/Homepage/Operations/RebalanceServer/RebalanceServerOptions.ts:
```diff
@@ -174,5 +174,35 @@ export const rebalanceServerOptions: RebalanceServerOption[] = [
+        "markWithWarningIcon": false
+    },
+    {
+        "name": "forceCommitBatchStatusCheckTimeoutMs",
```
wasn't there a 4th parameter that was added?

somandal commented on pinot-controller/src/main/resources/app/components/Homepage/Operations/RebalanceServer/RebalanceServerOptions.ts:
```diff
@@ -174,5 +174,35 @@ export const rebalanceServerOptions: RebalanceServerOption[] = [
+        "defaultValue": 180000,
+        "type": "INTEGER",
+        "label": "Force Commit Status Check Timeout (ms)",
+        "description": "If forceCommitBeforeMoved is set, this is the timeout in milliseconds for force commit batch status checks. Maximum time to wait for force commit operations to complete",
```
just make sure to rename the `forceCommitBeforeMoved` throughout this PR

somandal reviewed (commented):
still haven't had a chance for a thorough review, but wanted to get a few comments out

J-HowHuang commented:
Added manual test to check the force commit timeout case  in the PR description too

yashmayya commented on pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/TableRebalancePauselessIntegrationTest.java:
```diff
@@ -0,0 +1,409 @@
+    }, 1000L, timeoutMs, "Failed to complete rebalance");
+  }
+
+  private void waitForTableEVISConverge(String tableName, long timeoutMs) {
```
Unused?

yashmayya commented on pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/TableRebalancePauselessIntegrationTest.java:
```diff
@@ -0,0 +1,409 @@
+  }
+
+  @Test
+  public void testForceCommit()
```
Are these tests (nearly) identical to the ones added to `TableRebalanceIntegrationTest`?

J-HowHuang replied:
Yes, but I found this is the better way to set up a test for pauseless table? I'm not sure how to better modularize this

yashmayya reviewed (commented):
Thanks for the updates @J-HowHuang, this is looking pretty good! I've left a few more comments.

J-HowHuang commented on pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/TableRebalanceIntegrationTest.java:
```diff
@@ -1403,4 +1700,40 @@ private void waitForReloadToComplete(String reloadJobId, long timeoutMs) {
+        ServerRebalanceJobStatusResponse rebalanceStatus =
+            JsonUtils.stringToObject(httpResponse.getResponse(), ServerRebalanceJobStatusResponse.class);
+        return rebalanceStatus.getTableRebalanceProgressStats().getStatus() == RebalanceResult.Status.DONE;
+      } catch (Exception e) {
+        return null;
+      }
```
This is inside the `waitForCondition`'s condition function, so effectively it retries, until timeout.
The exceptions it catches `HttpErrorStatusException` and `JsonProcessingException`could be temporary (ZK not yet updated is what I thought of), so it's okay to just give it few more HTTP requests to try until timeout imo.

J-HowHuang commented on pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/TableRebalanceIntegrationTest.java:
```diff
@@ -1379,6 +1388,294 @@ public void testRebalanceJobZkMetadataCleanup()
+  }
+
+  @Test
+  void testForceCommitStrictReplicaGroup()
```
They're different in one step, where one test downtime and the other test don't, and test with batch size instead

J-HowHuang replied:
They are different in one step, where the other tests downtime rebalance and this does not. This one tests batch size instead.

yashmayya replied:
Yeah but it seems like most of the code is common, so I think it'd be good to refactor it to use [parameters](https://testng.org/parameters.html) and share the common code. It's okay for the different parts to simply be in an if-else block based on the parameter.

J-HowHuang replied:
Done

J-HowHuang commented on pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/TableRebalanceIntegrationTest.java:
```diff
@@ -1403,4 +1700,40 @@ private void waitForReloadToComplete(String reloadJobId, long timeoutMs) {
+        TableViews.TableView idealState =
+            JsonUtils.stringToObject(httpResponse.getResponse(), TableViews.TableView.class);
+
+        requestUrl = getControllerRequestURLBuilder().forIdealState(tableName);
```
Thank you good catch! Surprised this pass the test tho

yashmayya replied:
Well, we were only checking if IS = EV in this method right? And IS would always be equal to IS 😄

J-HowHuang commented on pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/TableRebalanceIntegrationTest.java:
```diff
@@ -1379,6 +1388,294 @@ public void testRebalanceJobZkMetadataCleanup()
+    serverStarter0.stop();
+    serverStarter1.stop();
+    serverStarter2.stop();
+    serverStarter3.stop();
```
done

J-HowHuang commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java:
```diff
@@ -673,6 +673,14 @@ public RebalanceResult rebalance(
+      @QueryParam("forceCommitBatchSize") int forceCommitBatchSize,
+      @ApiParam(value = "Interval in milliseconds for checking force commit batch status") @DefaultValue("5000")
+      @QueryParam("forceCommitBatchStatusCheckIntervalMs") int forceCommitBatchStatusCheckIntervalMs,
+      @ApiParam(value = "Timeout in milliseconds for force commit batch status check") @DefaultValue("180000")
```
Added the constants in `ForceCommitBatchConfig`, changed the reference in forceCommit API as well.

J-HowHuang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/ZkBasedTableRebalanceObserver.java:
```diff
@@ -151,6 +151,24 @@ public void onTrigger(Trigger trigger, Map<String, Map<String, String>> currentS
+        updatedStatsInZk = true;
+        break;
+      case FORCE_COMMIT_BEFORE_MOVED_END_TRIGGER:
+        LOGGER.info("force commit for consuming segments for table: {} is done",
```
It's in the method `forceCommitConsumingSegmentsAndWait`
https://github.com/apache/pinot/blob/419df3c246ccf6c5a1385faf564c00ce95dbd222/pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java#L2065-L2067

yashmayya commented on pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/TableRebalanceIntegrationTest.java:
```diff
@@ -1379,6 +1388,284 @@ public void testRebalanceJobZkMetadataCleanup()
+      // the sequence number should not increase since the consuming segment is not committed
+      assertEquals(consumingSegmentNow.getSequenceNumber(), consumingSegmentOriginal.getSequenceNumber());
+
+      // Resume the table
+      tableConfig.setTenantConfig(new TenantConfig(getBrokerTenant(), getServerTenant(), null));
+      tableConfig.getValidationConfig().setReplication("1");
+      updateTableConfig(tableConfig);
+      rebalanceConfig.setForceCommit(false);
+      rebalanceConfig.setMinAvailableReplicas(0);
+      rebalanceConfig.setDowntime(false);
+      rebalanceConfig.setIncludeConsuming(true);
+      response = sendPostRequest(getRebalanceUrl(rebalanceConfig, TableType.REALTIME));
+      rebalanceResult = JsonUtils.stringToObject(response, RebalanceResult.class);
+      waitForRebalanceToComplete(rebalanceResult.getJobId(), 30000);
```
IMO this should be in the finally block as well, since the other tests probably won't pass with the table in an unexpected tenant.

yashmayya commented on pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/TableRebalanceIntegrationTest.java:
```diff
@@ -1726,4 +1556,82 @@ private void waitForTableEVISConverge(String tableName, long timeoutMs) {
+   * Helper method to perform segment moving test with EVIS convergence wait.
+   * Similar to performSegmentMovingTest but waits for external view/ideal state convergence instead of rebalance completion.
+   */
+  void performSegmentMovingTestWithEVISConverge(RebalanceConfig rebalanceConfig, TableConfig tableConfig,
```
Unused?

yashmayya commented on pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/TableRebalanceIntegrationTest.java:
```diff
@@ -1726,4 +1556,82 @@ private void waitForTableEVISConverge(String tableName, long timeoutMs) {
+   * @param timeoutMs timeout in milliseconds
+   * @param waitForEVISConverge if true, waits for external view/ideal state convergence; if false, waits for rebalance completion
+   */
+  private void performSegmentMovingTest(RebalanceConfig rebalanceConfig, TableConfig tableConfig, String newTenant,
```
If this is identical to the one in the pauseless integration test class, I think we can move it to `BaseClusterIntegrationTestSet` and share it across both these classes.

J-HowHuang replied:
Done. Btw they share `BaseClusterIntegrationTest` instead of `BaseClusterIntegrationTestSet`

yashmayya replied:
We should update `BasePauselessRealtimeIngestionTest` to extend `BaseClusterIntegrationTestSet` instead of `BaseClusterIntegrationTest` and put this test over there. `BaseClusterIntegrationTest` has the test infra setup related methods, `BaseClusterIntegrationTestSet` is the one where shared / common tests go.

yashmayya commented on pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/TableRebalancePauselessIntegrationTest.java:
```diff
@@ -207,171 +205,129 @@ public void testForceCommit()
-          2);
-      assertEquals(summary.getSegmentInfo().getConsumingSegmentToBeMovedSummary().getNumConsumingSegmentsToBeMoved(),
-          4);
+  private void waitForRebalanceToComplete(String rebalanceJobId, long timeoutMs) {
```
Let's move these common utils to one of the base classes (probably `ControllerTest`) and re-use them instead of duplicating.

J-HowHuang replied:
Decided to put this in `BaseClusterIntegrationTest` since most of the non-integration tests have direct access to a `TableRebalancer` so they don't need this util

yashmayya reviewed (approved):
Thanks @J-HowHuang, LGTM! I just have a couple more minor comments about increasing code re-use across the new integration tests.

Jackie-Jiang commented:
@J-HowHuang The new added test is flaky. Can you please take a look?
Example run: https://github.com/apache/pinot/actions/runs/15924989688/job/44920115015?pr=16216
